core/arch/arm/emit_utils.c

```diff
@@ -980,10 +980,26 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
     add_patch_marker(patch, target_delete_entry, PATCH_ASSEMBLE_ABSOLUTE,
                      0 /* beginning of instruction */,
                      (ptr_uint_t *)&ibl_code->target_delete_entry);
+
     /* We just executed the hit path, so the app's r1 and r2 values are still in
      * their TLS slots, and &linkstub is still in the r3 slot.
      */
 
+    /* i#4665: We speculatively store next fragment tag in r2 (stored in dcontext later)
+     * and &linkstub_ibl_deleted in r1 (passed to fcache_return via r0 later). This can be
+     * tested on ibl-stress test. But that hasn't been ported to ARM yet, and we do not
+     * have an ARM machine available.
+     */
+    ASSERT_NOT_TESTED();
+    APP(&ilist,
+        INSTR_CREATE_ldr(
+            dc, opnd_create_reg(DR_REG_R2),
+            OPND_CREATE_MEMPTR(DR_REG_R1, offsetof(fragment_entry_t, tag_fragment))));
+    instrlist_insert_mov_immed_ptrsz(dc, (ptr_uint_t)get_ibl_deleted_linkstub(),
+                                     opnd_create_reg(DR_REG_R1), &ilist, NULL, NULL,
+                                     NULL);
+    APP(&ilist, instr_create_restore_from_tls(dc, DR_REG_R0, TLS_REG0_SLOT));
+
     /* Unlink path: entry from stub */
```

Comment:
For the `Unlink path` case, it seems that `fcache_return` will see the app's r1 value in r0, instead of the last linkstub address. (see L1011 -> 1017 -> 1019) 
@derekbruening

Reply:
Haven't made any fix for this yet.

Reply:
It looks correct I think.
Stub spills r1->r1_slot, r2->r2_slot; puts linkstub into r1, target into r2.
Unlink then spills r0->r0_slot, r1(linkstub) -> r3_slot.
Then on the shared miss path, it moves the linkstub from r3_slot back to r1.
Then it moves the linkstub from r1 into r0.
Then it restores r1_slot->r1, moves target from r2 into ntext_tag, restores r2.

Reply:
Oh you don't mean entering at the unlink entry point: you mean taking the target delete path and falling through?

Reply:
Oh right. `r1` at this point has the linkstub already, not the app value. (Maybe I kept thinking `r0` because of AArch64). Thanks.

Reply:
For the target delete path and falling through, I already made a fix above, by loading `linkstub_ibl_deleted` in r1.